### PR TITLE
Convert Perl lexer to new format

### PR DESCRIPTION
**Pull request by snoopy**
I've converted the Perl lexer to the new format. Based on my quick testing it seems to work fine.
Could you please have a look as well?

### CODE REVIEW

**Comment by orbitalquark** [comment]
Thank you! I will take a look when I have some time. Sorry for the delay.

File: lexers/perl.lua [hunk 1]
```diff
@@ -1,52 +1,16 @@
...
-  'rename', 'reset', 'reverse', 'rewinddir', 'rindex', 'rmdir', 'scalar', 'seekdir', 'seek',
-  'select', 'semctl', 'semget', 'semop', 'send', 'setgrent', 'sethostent', 'setnetent', 'setpgrp',
-  'setpriority', 'setprotoent', 'setpwent', 'setservent', 'setsockopt', 'shift', 'shmctl', 'shmget',
-  'shmread', 'shmwrite', 'shutdown', 'sin', 'sleep', 'socket', 'socketpair', 'sort', 'splice',
-  'split', 'sprintf', 'sqrt', 'srand', 'stat', 'study', 'substr', 'symlink', 'syscall', 'sysread',
-  'sysseek', 'system', 'syswrite', 'telldir', 'tell', 'tied', 'tie', 'time', 'times', 'truncate',
-  'ucfirst', 'uc', 'umask', 'undef', 'unlink', 'unpack', 'unshift', 'untie', 'utime', 'values',
-  'vec', 'wait', 'waitpid', 'wantarray', 'warn', 'write'
-}))
+lex:add_rule('marker', lex:tag(lexer.COMMENT, lex:word_match(lexer.COMMENT) * lexer.any^0))
```
Comment (orbitalquark) [inline lexers/perl.lua]: I would switch back to `lexer.word_match()` here because comment words are quite static.

File: lexers/perl.lua [hunk 2]
```diff
@@ -102,25 +66,31 @@ local lit_str = 'q' * P('q')^-1 * literal_delimited
...
   sq_str + dq_str + cmd_str + heredoc + lit_str + lit_array + lit_cmd + lit_tr)
 local regex_str = lexer.after_set('-<>+*!~\\=%&|^?:;([{', lexer.range('/', true) * S('imosx')^0)
 local lit_regex = 'qr' * literal_delimited * S('imosx')^0
 local lit_match = 'm' * literal_delimited * S('cgimosx')^0
 local lit_sub = 's' * literal_delimited2 * S('ecgimosx')^0
-local regex = token(lexer.REGEX, regex_str + lit_regex + lit_match + lit_sub)
+local regex = lex:tag(lexer.REGEX, regex_str + lit_regex + lit_match + lit_sub)
 lex:add_rule('string', string + regex)
 
+-- Functions.
```
Comment (orbitalquark) [inline lexers/perl.lua:78]: What was the reason for switching the rule order here? Is there something about strings that prevents them from matching after functions/methods? (I do not know much about Perl.)

**Review by orbitalquark — CHANGES_REQUESTED** [review]
This looks good for the most part. Thank you for taking the time to migrate this lexer! Does Perl have any built-in constants worth adding? This would be a good opportunity to do so. Otherwise, no problem.

File: lexers/perl.lua (hunk 2 above)
Comment (snoopy) [reply lexers/perl.lua:78]: During testing the old order broke regex strings. The easiest fix was adjusting the rule order.
Perl being Perl it still might break in unexpected ways.

**Comment by snoopy** [comment]
> [...] Does Perl have any built-in constants worth adding? This would be a good opportunity to do so. Otherwise, no problem.

Not that I know of unless you count special vars (https://perldoc.perl.org/variables). They are not really constant though.

**Comment by orbitalquark** [comment]
Can we use `lexer.VARIABLE_BUILTIN` for those? I think that will be useful. Maybe not all of them, but some useful subset, mainly so that the word list exists and other users can extend it to their liking.

File: lexers/perl.lua (hunk 1 above)
Comment (snoopy) [reply lexers/perl.lua]: You mean like this? 
```lua
lex:add_rule('marker', lex:tag(lexer.COMMENT, lex:word_match('__DATA__ __END__')))

Comment (orbitalquark) [reply lexers/perl.lua]: No, just straight up `lexer.word_match('__DATA__ __END__')`. The updated JSON lexer uses it: https://github.com/orbitalquark/scintillua/blob/aaabb16898389b0e7a130ca5f31db420cfa4164a/lexers/json.lua#L16

Comment (snoopy) [reply lexers/perl.lua]:
```lua
lex:add_rule('marker', lex:tag(lexer.KEYWORD, lexer.word_match('__DATA__ __END__')))
```

This actually does not style everything after `__DATA__` as comment which might be nicer.

Comment (orbitalquark) [reply lexers/perl.lua]: It would be nicer, but rules that span multiple lines are quite brittle in my experience.

**Comment by snoopy** [comment]
Variables are a bit spotty. LMK if there are any obvious mistakes.

**Comment by snoopy** [comment]
Something is still wrong with the function/string rules. I'll try to figure it out.

**Comment by snoopy** [comment]
Functions should work better now. The variables are still having problems where some of them don't get detected properly. The rules are probably not quite right.

What kind of changes should be made to the marker rule?

**Comment by orbitalquark** [comment]
The marker rule should still mark `__DATA__` or `__END__` followed by anything after as a comment, just like before.

File: lexers/perl.lua [hunk 3]
```diff
@@ -132,15 +102,143 @@ local special_var = '$' * (
   lexer.digit^1)
 -- LuaFormatter on
 local plain_var = ('$#' + S('$@%')) * P('$')^0 * lexer.word + '$#'
-lex:add_rule('variable', token(lexer.VARIABLE, special_var + plain_var))
+lex:add_rule('variable_builtin', lex:tag(lexer.VARIABLE_BUILTIN, lex:word_match(lexer.VARIABLE_BUILTIN)))
```
Comment (orbitalquark) [inline lexers/perl.lua]: Built-in variables should not all be grouped into a single word match. Scintillua looks for the longest sequence of characters matching letters and all special characters used inside a word list and then does the lookup. This is probably the cause of not detecting some variables properly.

Instead, something like `plain_var` above should be done for grouping variables into subsets. For example, ``'$' * (lpeg.R('19') + S('!"$%&\'()+,-./:;<=>?@\\]_`|~') + '^' * S('ACDEFHILMNOPRSTVWX')^-1 + 'ARGV'`` will cover most of the '$' variables. You can do something similar for '%' and '@' variables.

**Comment by snoopy** [comment]
Thanks for the pointers. Seems to work fine now.

**Comment by orbitalquark** [comment]
Thank you! Committed with some minor formatting changes via https://github.com/orbitalquark/scintillua/commit/ce13dedf4a0d7427e3063271d7c8ca89c0530f16

**Pull request closed without merging**